File: code/001_climate_and_soil_data_alignment.R
#######################################################################
# Evangeline Corcoran,  Version: 21/08/2024
# Retrieve climate data matching land use and soil data grid points
#######################################################################
install.packages("raster")
install.packages("rgdal")
install.packages("dplyr")
install.packages("Rcpp")
install.packages("ncdf4")
install.packages('maptools')

#load packages
library(Rcpp)
library(raster)
library(rgdal)
library(maptools)
library(dplyr)
library(ncdf4)

setwd('~/Large Scale Model Data')
#read in grid points
grid <- read.csv('LargeScaleCropData_grid_points.csv')

#read in met data by variable
setwd('~/Large Scale Model Data/met/2017/peti')
stack_1 <- stack("chess-pe_peti_gb_1km_daily_20170101-20170131.nc")
stack_2 <- stack("chess-pe_peti_gb_1km_daily_20170201-20170228.nc")
stack_3 <- stack("chess-pe_peti_gb_1km_daily_20170301-20170331.nc")
stack_4 <- stack("chess-pe_peti_gb_1km_daily_20170401-20170430.nc")
stack_5 <- stack("chess-pe_peti_gb_1km_daily_20170501-20170531.nc")
stack_6 <- stack("chess-pe_peti_gb_1km_daily_20170601-20170630.nc")
stack_7 <- stack("chess-pe_peti_gb_1km_daily_20170701-20170731.nc")
stack_8 <- stack("chess-pe_peti_gb_1km_daily_20170801-20170831.nc")
stack_9 <- stack("chess-pe_peti_gb_1km_daily_20170901-20170930.nc")
stack_10 <- stack("chess-pe_peti_gb_1km_daily_20171001-20171031.nc")
stack_11 <- stack("chess-pe_peti_gb_1km_daily_20171101-20171130.nc")
stack_12 <- stack("chess-pe_peti_gb_1km_daily_20171201-20171231.nc")

#convert to dataframes
df_1 <- as.data.frame(stack_1, xy= TRUE)
df_2 <- as.data.frame(stack_2, xy= TRUE)
df_3 <- as.data.frame(stack_3, xy= TRUE)
df_4 <- as.data.frame(stack_4, xy= TRUE)
df_5 <- as.data.frame(stack_5, xy= TRUE)
df_6 <- as.data.frame(stack_6, xy= TRUE)
df_7 <- as.data.frame(stack_7, xy= TRUE)
df_8 <- as.data.frame(stack_8, xy= TRUE)
df_9 <- as.data.frame(stack_9, xy= TRUE)
df_10 <- as.data.frame(stack_10, xy= TRUE)
df_11 <- as.data.frame(stack_11, xy= TRUE)
df_12 <- as.data.frame(stack_12, xy= TRUE)

#merge all variable data for year
merge_df <- merge(df_1, df_2, by = c("x","y"))
merge_df <- merge(merge_df, df_3, by = c("x","y"))
merge_df <- merge(merge_df, df_4, by = c("x","y"))
merge_df <- merge(merge_df, df_5, by = c("x","y"))
merge_df <- merge(merge_df, df_6, by = c("x","y"))
merge_df <- merge(merge_df, df_7, by = c("x","y"))
merge_df <- merge(merge_df, df_8, by = c("x","y"))
merge_df <- merge(merge_df, df_9, by = c("x","y"))
merge_df <- merge(merge_df, df_10, by = c("x","y"))
merge_df <- merge(merge_df, df_11, by = c("x","y"))
merge_df <- merge(merge_df, df_12, by = c("x","y"))

# FROM (BNG) TO (WGS84)
bng <- "+init=epsg:27700"
wgs84 <- "+init=epsg:4326"

# Create spatial object where eastings is x-axis and northings is y-axis
coordinates(merge_df) <- c("x", "y")
# Set projection of data
proj4string(merge_df) <- CRS(bng)

# BNG to WGS84 conversion
merge_wgs84 = spTransform(merge_df, CRS(wgs84))

# Convert SpatialPoints back into a data-frame
merge_wgs84 <- as.data.frame(merge_wgs84)

#round coordinates in grid and met data to 7 significant figures
merge_wgs84$x <- signif(merge_wgs84$x, digits = 7)
merge_wgs84$y <- signif(merge_wgs84$y, digits = 7)
grid$x <- signif(grid$x, digits = 7)
grid$y <- signif(grid$y, digits = 7)

#filter met data for grid points matching land use and soil
merge_grid <- merge(merge_wgs84, grid, by = c("x","y"))

#remove missing data
merge_grid <- na.omit(merge_grid)

#save data
write.csv(merge_grid, 'LargeScaleCropData_met_peti_2017.csv')

#retrieve and save updated grid points
setwd('~/Large Scale Model Data')
grids <- as.data.frame(cbind(merge_grid$x, merge_grid$y))
grids <- grids %>% rename(x = V1,
                          y = V2)
write.csv(grids, 'LargeScaleCropData_grid_points.csv')

File: code/001_climate_and_soil_data_alignment.py
#######################################################################
# Evangeline Corcoran
# Version: 21/08/2024
#
# Retrieve climate data matching land use and soil data grid points
# Modernised: July 2026
#######################################################################

# pip install pandas numpy xarray netCDF4 pyproj

import os
import numpy as np
import pandas as pd
import xarray as xr
from pyproj import Transformer

############################
# Project paths
############################

# project root is one level up from this script's folder (code/), data
# lives in project_root/data
code_dir = os.path.dirname(os.path.abspath(__file__))
project_root = os.path.dirname(code_dir)
data_dir = project_root + "/data/"
met = data_dir + "met/2017/peti/"

############################
# Helpers
############################

# round to n significant figures
def signif(x, digits):
    x = np.asarray(x, dtype = float)
    out = np.zeros_like(x)
    nz = x != 0
    mags = np.floor(np.log10(np.abs(x[nz])))
    factor = 10 ** (digits - 1 - mags)
    out[nz] = np.round(x[nz] * factor) / factor
    return out

############################
# Read grid points
############################

grid = pd.read_csv(data_dir + "LargeScaleCropData_grid_points.csv")

############################
# Read NetCDF files
############################

file_names = [
    "chess-pe_peti_gb_1km_daily_20170101-20170131.nc",
    "chess-pe_peti_gb_1km_daily_20170201-20170228.nc",
    "chess-pe_peti_gb_1km_daily_20170301-20170331.nc",
    "chess-pe_peti_gb_1km_daily_20170401-20170430.nc",
    "chess-pe_peti_gb_1km_daily_20170501-20170531.nc",
    "chess-pe_peti_gb_1km_daily_20170601-20170630.nc",
    "chess-pe_peti_gb_1km_daily_20170701-20170731.nc",
    "chess-pe_peti_gb_1km_daily_20170801-20170831.nc",
    "chess-pe_peti_gb_1km_daily_20170901-20170930.nc",
    "chess-pe_peti_gb_1km_daily_20171001-20171031.nc",
    "chess-pe_peti_gb_1km_daily_20171101-20171130.nc",
    "chess-pe_peti_gb_1km_daily_20171201-20171231.nc",
]

############################
# Convert to data frames
############################

# one row per (x, y) cell, one column per day. day columns named
# X2017.01.01 style so downstream matching on the X-prefix works
dfs = []
for f in file_names:
    ds = xr.open_dataset(met + f)
    tdf = ds["peti"].to_dataframe().reset_index()
    wide = tdf.pivot(index = ["x", "y"], columns = "time", values = "peti")
    wide.columns = ["X" + pd.Timestamp(c).strftime("%Y.%m.%d") for c in wide.columns]
    dfs.append(wide.reset_index())
    ds.close()

############################
# Merge monthly data
############################

merge_df = dfs[0]
for tdf in dfs[1:]:
    merge_df = merge_df.merge(tdf, on = ["x", "y"])

############################
# Convert BNG to WGS84
############################

transformer = Transformer.from_crs("epsg:27700", "epsg:4326", always_xy = True)
lon, lat = transformer.transform(merge_df["x"].to_numpy(), merge_df["y"].to_numpy())
merge_wgs84 = merge_df.copy()
merge_wgs84["x"] = lon
merge_wgs84["y"] = lat

############################
# Round coordinates
############################

merge_wgs84["x"] = signif(merge_wgs84["x"], digits = 7)
merge_wgs84["y"] = signif(merge_wgs84["y"], digits = 7)
grid["x"] = signif(grid["x"], digits = 7)
grid["y"] = signif(grid["y"], digits = 7)

############################
# Merge climate and grid data
############################

merge_grid = merge_wgs84.merge(grid, on = ["x", "y"])

# sort by the join columns so the ordering is deterministic
merge_grid = merge_grid.sort_values(["x", "y"])

# remove missing data
merge_grid = merge_grid.dropna()

############################
# Save outputs
############################

merge_grid.to_csv(data_dir + "LargeScaleCropData_met_peti_2017.csv", index = False)

# overwrites the input grid file with the surviving grid points
grids = merge_grid[["x", "y"]].copy()
grids.to_csv(data_dir + "LargeScaleCropData_grid_points.csv", index = False)
